crates/components/authdecode/authdecode-core/src/backend/halo2/circuit.rs

```diff
@@ -0,0 +1,606 @@
+use halo2_poseidon::poseidon::{primitives::ConstantLength, Hash, Pow5Chip, Pow5Config};
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter, Region, SimpleFloorPlanner, Value},
+    halo2curves::bn256::Fr as F,
+    plonk::{
+        Advice, Circuit, Column, ConstraintSystem, Constraints, Error, Expression, Instance,
+        Selector,
+    },
+    poly::Rotation,
+};
+use std::convert::TryInto;
+
+use crate::backend::halo2::{
+    poseidon::{configure_poseidon_rate_15, configure_poseidon_rate_2},
+    utils::{compose_bits, f_to_bits},
+};
+
+use poseidon_halo2::{Spec15, Spec2};
+
+// Rationale for the selection of constants.
+//
+// In order to optimize the proof generation time, the circuit should contain as few instance
+// columns as possible and also as few rows as possible.
+// The circuit has [super::CHUNK_SIZE] public inputs (deltas) which must be placed into instance
+// columns. It was empirically established that 58 rows and 64 instance columns provides the best
+// performance.
+//
+// Note that 58 usable rows is what we get when we set the circuit's K to 6 (halo2 reserves 6 rows
+// for internal purposes, so we get 2^6-6 usable rows).
+
+/// How many field elements to use to pack the plaintext into. Only [USABLE_BYTES] of each field
+/// element will be used.  
+pub const FIELD_ELEMENTS: usize = 14;
+
+/// How many least significant bytes of a field element to use to pack the plaintext into.
+pub const USABLE_BYTES: usize = 31;
```

Comment:
what's the reason we dont use all 32 bytes here?

Reply:
because the field size is slightly less that 256 bits.